### ISSUE TITLE
Foreign keys: Updating co-located placements in different order could cause deadlock

### ISSUE THREAD

**Issue by marcocitus** [citusdata/citus#861]
If shard placements have foreign keys and the replication factor is >1, then we need to make sure we always perform commands in the same order across *all co-located placements*. Otherwise, a write to the referenced table could take locks that will block the writes on the referring table and create a deadlock.

A way to simulate the deadlock in PostgreSQL below.

```
CREATE TABLE main_1 (x int, y int);
CREATE TABLE main_2 (x int, y int);
CREATE TABLE reference_1 (a int unique, b int);
CREATE TABLE reference_2 (a int unique, b int);
ALTER TABLE main_1 ADD CONSTRAINT main_1_cs FOREIGN KEY (x) REFERENCES reference_2 (a);
ALTER TABLE main_2 ADD CONSTRAINT main_2_cs FOREIGN KEY (x) REFERENCES reference_2 (a);
INSERT INTO reference_1 VALUES (1,1);
INSERT INTO reference_2 VALUES (1,1);
```

In session 1, write to first shard placement of main table:
```
BEGIN;
INSERT INTO main_1 VALUES (1,1);
```

In session 2, write to second shard placement of referenced table:
```
BEGIN;
DELETE FROM reference_2 WHERE a = 1;
```

In session 1, write to second shard placement of main table.
```
INSERT INTO main_2 VALUES (1,1);
...blocks waiting on DELETE to finish ...
```

In session 2, write to first shard placement of referenced table.
```
DELETE FROM reference_1 WHERE a = 1;
ERROR:  deadlock detected
DETAIL:  Process 11755 waits for ShareLock on transaction 2677; blocked by process 11751.
Process 11751 waits for ShareLock on transaction 2678; blocked by process 11755.
HINT:  See server log for query details.
CONTEXT:  while deleting tuple (0,1) in relation "reference_1"
```

In the Citus case, the deadlock wouldn't actually be detected since it's happening across 2 machines.

To solve this, we either need very aggressive write locks, or always write to nodes in the same order (e.g. ordered by node ID).

**Comment by onderkalaci** [comment]
Relates to #698

**Comment by ozgune** [comment]
Quick question. Did we reach a decision on whether we'd like to fix this issue?

**Comment by ozgune** [comment]
We talked about this in an internal email thread, and I'm adding the email thread to here.

--
IIRC, we sent our writes to shard placements in a deterministic order. That is, if a shard has two placements, we always first send our writes to placement-a and then to placement-b.

I created two tables main and reference, and distributed them by hash on (a) and (x) respectively. I then inserted values into main and deleted from reference in the orders we laid out in the issue, and several different variations, through two separate psql prompts.

I couldn't reproduce the issue but I didn't spend a lot of time on it either. I had two questions:

* For this scenario to happen, does the user need to change the ``all_modifications_commutative`` flag?
* Can we easily reproduce this deadlock in the community edition? Or do we have a particular scenario where we trigger the deadlocks?

**Comment by marcocitus** [comment]
> IIRC, we sent our writes to shard placements in a deterministic order.

Yes, currently in placement ID order, but this should probably be node ID order since placement ID changes after a shard move and may thus follow a different order.

> For this scenario to happen, does the user need to change the all_modifications_commutative flag?

No, the deadlock is on row-level locks, not shard resource locks.

> Can we easily reproduce this deadlock in the community edition? Or do we have a particular scenario where we trigger the deadlocks?

No, we don't currently support foreign keys for replication factor >1.

**Comment by onderkalaci** [comment]
Just a note while trying it:
> I couldn't reproduce the issue but I didn't spend a lot of time on it either

There seems to be a typo in the following line (i..e, referenced table should be `reference_1`)

> ALTER TABLE main_1 ADD CONSTRAINT main_1_cs FOREIGN KEY (x) REFERENCES reference_2 (a);


Should be:

ALTER TABLE main_1 ADD CONSTRAINT main_1_cs FOREIGN KEY (x) REFERENCES reference_1 (a);

**Comment by onderkalaci** [comment]
Is this issue still relevant for us even if we implement fkey to reference tables? There seems to be an actual deadlock in this case and Citus would detect it. 

@marcocitus do you think is this still an issue for Citus?
